src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs

```diff
@@ -642,26 +642,44 @@ private bool ReplacementBreaksQueryClause(QueryClauseSyntax originalClause, Quer
                 !SymbolInfosAreCompatible(originalClauseInfo.OperationInfo, newClauseInfo.OperationInfo);
         }
 
+
         protected override bool ConversionsAreCompatible(SemanticModel originalModel, ExpressionSyntax originalExpression, SemanticModel newModel, ExpressionSyntax newExpression)
         {
-            if (originalModel == null || originalExpression == null || newModel == null || newExpression == null)
-            {
-                return false;
-            }
-
             return ConversionsAreCompatible(originalModel.GetConversion(originalExpression), newModel.GetConversion(newExpression));
         }
 
         protected override bool ConversionsAreCompatible(ExpressionSyntax originalExpression, ITypeSymbol originalTargetType, ExpressionSyntax newExpression, ITypeSymbol newTargetType)
         {
-            if (originalExpression == null || originalTargetType == null || newExpression == null || newTargetType == null)
+            Conversion? originalConversion = null;
+            Conversion? newConversion = null;
+
+            if (HasNonNullType(this.OriginalSemanticModel, originalExpression) &&
+                HasNonNullType(this.SpeculativeSemanticModel, newExpression))
             {
-                return false;
+                originalConversion = this.OriginalSemanticModel.ClassifyConversion(originalExpression, originalTargetType);
+                newConversion = this.SpeculativeSemanticModel.ClassifyConversion(newExpression, newTargetType);
+            }
+            else
+            {
+                ITypeSymbol originalConvertedTypeSymbol = null;
+                if (TryGetConvertedTypeForExpression(this.OriginalSemanticModel, originalExpression, out originalConvertedTypeSymbol))
+                {
+                    originalConversion = this.OriginalSemanticModel.Compilation.ClassifyConversion(originalConvertedTypeSymbol, originalTargetType);
```

Comment:
All this code (along with the above code in the ```if``` portoin) seems ripe for some sort of cleanup.  Aren't you just basically saying:

```
var typeINfo = model.GetTypeInfo(expression)
var type = typeInfo.Type ?? typeInfo.ConvertedType
ver conversion = model.ClassifyConvertion(type, originalTargetType)
```